### PR TITLE
Add error log for files with non-unicode symbol 

### PR DESCRIPTION
**Pull request by VitaliyaIoffe**
If a temporary test result file contains a non-unicode symbol, it will raise
additional logging and will fail the test.

Closes: #293

### CODE REVIEW

File: lib/pytap13.py [hunk 1]
```diff
@@ -17,43 +17,37 @@
...
-    from StringIO import StringIO
-except ImportError:
-    # Python 3
-    from io import StringIO
+from io import BytesIO
 
 import yaml
-from lib.utils import string_types
 
+from lib.utils import string_types, binary_regex, str_to_bytes, bytes_to_str
```
Comment (LeonidVas) [inline lib/pytap13.py:24]: Use a separate line for each import (I don't remember where these rules come from, but we follow it).

File: lib/pytap13.py [hunk 2]
```diff
@@ -17,43 +17,37 @@
...
 
-RE_VERSION = re.compile(r"^\s*TAP version 13\s*$")
-RE_PLAN = re.compile(
+RE_VERSION = binary_regex(r"^\s*TAP version 13\s*$")
```
Comment (LeonidVas) [inline lib/pytap13.py:26]: Why not just `re.compile(b."....")`?

File: lib/pytap13.py [hunk 3]
```diff
@@ -102,8 +96,8 @@ def _parse(self, source):
                     in_yaml = True
                     continue
 
-            on_top_level = not line.startswith('    ')
-            raw_line = line.rstrip('\n')
+            on_top_level = not line.startswith(str_to_bytes('    '))
```
Comment (LeonidVas) [inline lib/pytap13.py:99]: Why not just `(b'....')`? Here and below.

File: lib/test.py [hunk 4]
```diff
@@ -9,15 +8,17 @@
 from functools import partial
 from hashlib import md5
 
+import gevent
```
Comment (LeonidVas) [inline lib/test.py:11]: From https://github.com/tarantool/tarantool/wiki/Code-review-procedure#general-coding-points-to-check : "Don't do a change if it is not related to the patch."
Shuffling the import order is not related to the problem solved in the patch. Here and below.

File: lib/utils.py [hunk 5]
```diff
@@ -351,3 +355,8 @@ def str_to_bytes(s):
     if PY2:
         return s
     return s.encode('utf-8')
+
+
+def binary_regex(s, flags=0):
```
Comment (LeonidVas) [inline lib/utils.py:360]: I don't think this function is really useful for two reasons:
1) It adds 3 unnecessary function calls (`binary_regex`, `str_to_bytes` and `encode`) and doesn't reduce the amount of code.
2) It's so simple, just write `re.compile (b" ")`.

**Review by LeonidVas — CHANGES_REQUESTED** [review]
Hi! Thank you for the patch.
About the commit title:
Maybe something like: "Switch to using binary data in tests" is better? And I propose to move the description of the problem with decoding to the commit message.
About the commit message:
- From https://github.com/tarantool/tarantool/wiki/Code-review-procedure#commit-message : "Commit message should fit 72 symbols line width."
- "Fixes: #293"

Have you check that it works fine in both python2 and python3?

Also see several comments bellow:

File: lib/pytap13.py (hunk 2 above)
Comment (VitaliyaIoffe) [reply lib/pytap13.py:26]: Binary strings with the prefix "b" are allowed only in python3. We discussed this approach with additional function with @Totktonada

File: lib/utils.py (hunk 5 above)
Comment (VitaliyaIoffe) [reply lib/utils.py:360]: As I said above it is not so easy due to the "rb" prefix. Our patterns are raw strings. Either we should avoid raw strings and use b-prefix, or use this method.

File: lib/test.py (hunk 4 above)
Comment (VitaliyaIoffe) [reply lib/test.py:11]: Got it
Thank you

File: lib/pytap13.py (hunk 3 above)
Comment (VitaliyaIoffe) [reply lib/pytap13.py:99]: Maybe I didn't understand @Totktonada correctly. He said that there was a discussion about strings in the test-run and how to use them to support both python's versions. Consequently, you don't use literals associated with unicode and binary strings. So, I used str_to_bytes function. I was wrong?

File: lib/pytap13.py (hunk 1 above)
Comment (VitaliyaIoffe) [reply lib/pytap13.py:24]: Ok. 
Thank you

File: lib/pytap13.py (hunk 2 above)
Comment (VitaliyaIoffe) [reply lib/pytap13.py]: There is "br" literal in python 2.7. Fixed.

File: lib/pytap13.py (hunk 3 above)
Comment (VitaliyaIoffe) [reply lib/pytap13.py]: Fixed

**Comment by Totktonada** [comment]
(Related to several discussions above.) `str_to_bytes()` and `bytes_to_str()` are general purpose functions. If we can make the code easier for the byte array literal case and don't use those functions here, it looks okay for me, at least on the first glance.

**Comment by VitaliyaIoffe** [comment]
Yes, I've checked it with python2 and python3

**Comment by Totktonada** [comment]
Please, drop `broken_unicode.result`: so the change will be tested (see https://github.com/tarantool/test-run/issues/293#issuecomment-819601782). Or, better, add another similar test without a result file.

File: lib/pytap13.py [hunk 6]
```diff
@@ -174,13 +169,13 @@ def _parse(self, source):
         if len(self.tests) != self.tests_planned:
             comment = 'DIAG: Expected %s tests, got %s' % \
                 (self.tests_planned, len(self.tests))
-            self.tests.append(Test('not ok', len(self.tests), comment=comment))
+            self.tests.append(Test(b'not ok', len(self.tests), comment=comment))
 
     def parse(self, source):
         if isinstance(source, string_types):
-            self._parse(StringIO(source))
+            self._parse(BytesIO(source))
```
Comment (Totktonada) [inline lib/pytap13.py:176]: (Not a real review comment, just note. I'll give more thoughs a bit later.)

`string_types` is `bytes` and `unicode` on Python 2 and just `str` (unicode string) on Python 3. Let's consider Python 3 case. BytesIO accepts only `bytes`, not `str`:

```python
>>> from io import BytesIO
>>> BytesIO('')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: a bytes-like object is required, not 'str'
```

So the if-condition (`isinstance(source, string_types)`) contradicts with the `BytesIO(source)` expression.

**Comment by Totktonada** [comment]
(Sorry, I'm not good in concise explanations, but I'm always in attempts to do this better...)

There are several ideas that I keep in the mind here:

1. It is nice, when things just work, without artificial limitations. So, it would be good to support non-utf-8 data in TAP 13 results.
2. OTOH, complexity (maintainability) of the implementation matters. I think we should keep pieces of code that work with binary data quite dense and don't spread them much across APIs. (See 395edeb6b743c4479a62dd2183062124973d2b2a about the idea.)

To be honest, I don't know what we should choose here. I'll share several more thoughts around and will take timeout to think more.

* The `Test` class contains both binary and string data after the patch. It does not look nice. Moreover, the class is used outside of pytap13.py that descreases locality of the binary data.
* The pytap13.py code mixes byte literals and string literals after the patch and it looks error prone. It is not easy to verify correctness. Let's consider line 152 (157 on `master`): `Test('not ok', <...>)` — it'll raise `ValueError` from `str_to_bytes()`.
* If we'll choose to support only UTF-8 TAP 13 results, we should give a good error message instead of traceback and `UnicodeDecodeError`. It'll push a test author to either keep a result file or to make test cases description human readable. Not so bad.
* I had a thought to encode non-UTF-8 data in a test case description automatically in the `tap` built-in module, but it does not follow perl's Test::More behaviour.
* TAP 13 may contain YAML blocks and our YAML encoder uses `!!binary` tag with base64 encoding to represent non-printable or non-utf-8 characters. So the only piece of data that may break UTF-8 encoding is a test case description.

There are several ways to more forward:

* Decide to don't support non-UTF-8 data in a TAP 13 output and give a meaningful error message instead.
* Decide to support non-UTF-8 data in a TAP 13 output, but rearrange TAP 13 related code to don't spread byte arrays much.
* Proceed with a patch like current one, but elaborate agreemtn how the code should be refactored in a future.

File: lib/pytap13.py (hunk 6 above)
Comment (Totktonada) [reply lib/pytap13.py:176]: (In fact, after the patch, we never going into this branch, so it becomes a dead code.)

**Comment by LeonidVas** [comment]
> (Sorry, I'm not good in concise explanations, but I'm always in attempts to do this better...)
> 
> There are several ideas that I keep in the mind here:
> 
> 1. It is nice, when things just work, without artificial limitations. So, it would be good to support non-utf-8 data in TAP 13 results.
> 2. OTOH, complexity (maintainability) of the implementation matters. I think we should keep pieces of code that work with binary data quite dense and don't spread them much across APIs. (See [395edeb](https://github.com/tarantool/test-run/commit/395edeb6b743c4479a62dd2183062124973d2b2a) about the idea.)
> 
> To be honest, I don't know what we should choose here. I'll share several more thoughts around and will take timeout to think more.
> 
> * The `Test` class contains both binary and string data after the patch. It does not look nice. Moreover, the class is used outside of pytap13.py that descreases locality of the binary data.
> * The pytap13.py code mixes byte literals and string literals after the patch and it looks error prone. It is not easy to verify correctness. Let's consider line 152 (157 on `master`): `Test('not ok', <...>)` — it'll raise `ValueError` from `str_to_bytes()`.
> * If we'll choose to support only UTF-8 TAP 13 results, we should give a good error message instead of traceback and `UnicodeDecodeError`. It'll push a test author to either keep a result file or to make test cases description human readable. Not so bad.
> * I had a thought to encode non-UTF-8 data in a test case description automatically in the `tap` built-in module, but it does not follow perl's Test::More behaviour.
> * TAP 13 may contain YAML blocks and our YAML encoder uses `!!binary` tag with base64 encoding to represent non-printable or non-utf-8 characters. So the only piece of data that may break UTF-8 encoding is a test case description.
> 
> There are several ways to more forward:
> 
> * Decide to don't support non-UTF-8 data in a TAP 13 output and give a meaningful error message instead.
> * Decide to support non-UTF-8 data in a TAP 13 output, but rearrange TAP 13 related code to don't spread byte arrays much.
> * Proceed with a patch like current one, but elaborate agreemtn how the code should be refactored in a future.

Just raw thoughts. If we decide to use base64, it will hide where the error was in the raw data (which byte was wrong). Although it may not matter.

File: lib/test.py [hunk 7]
```diff
@@ -341,8 +341,15 @@ def tap_parse_print_yaml(self, yml):
 
     def check_tap_output(self):
         """ Returns is_tap, is_ok """
-        with open(self.tmp_result, 'r') as f:
-            content = f.read()
+        try:
+            with open(self.tmp_result, 'r') as f:
+                content = f.read()
+        except UnicodeDecodeError as e:
+            color_stdout('\nCannot read result file (%s).\n' % str(e),
```
Comment (LeonidVas) [inline lib/test.py:348]: It might be nice to write something about “maybe you are using binary” or “try to use base64” ... Up to you.

**Review by LeonidVas — COMMENTED** [review]
Hi! Thank you for the patch.
LGTM.

**Review by LeonidVas — APPROVED** [review]

File: lib/test.py (hunk 7 above)
Comment (Totktonada) [reply lib/test.py:348]: The actual message looks so:

> Cannot read result file ('utf-8' codec can't decode byte 0xc2 in position 27: invalid continuation byte).

However I would highlight that we only unable to use it as TAP13: file-to-file comparison still work. The 'TAP13 parse failed' message below looks good here (and also suggest `--update-result`). So, maybe, just move open+read under the try-except block and catch UnicodeDecodeError in the except clause?

Comment (Totktonada) [reply lib/test.py:348]: BTW, it would be good to highlight the fact that the change is only about parsing output as TAP13 somewhere in the commit message. Now it is misleading a bit.

Comment (VitaliyaIoffe) [reply lib/test.py]: @Totktonada fixed

File: lib/pytap13.py (hunk 6 above)
Comment (Totktonada) [reply lib/pytap13.py]: Since we decided to raise a good error instead of support of non UTF-8 data in TAP13 parser, this discussion is not actual anymore.

File: lib/test.py (hunk 7 above)
Comment (Totktonada) [reply lib/test.py]: Thanks!

**Review by Totktonada — APPROVED** [review]
LGTM.

**Comment by Totktonada** [comment]
Updated the test-run submodule in tarantool in [2.9.0-104-g7fa70b661](https://github.com/tarantool/tarantool/commit/7fa70b6619be693db3831411e306c6a62c35e655), 2.8.1-75-g029d66355, 2.7.2-62-g6f9ccad83, 1.10.10-25-g9b4acec8f.
